src/wallet/wallet.cpp

```diff
@@ -330,6 +338,36 @@ bool CWallet::WriteKeyMetadata(const CKeyMetadata& meta, const CPubKey& pubkey,
     return WalletBatch(*database).WriteKeyMetadata(meta, pubkey, overwrite);
 }
 
+void CWallet::UpgradeKeyMetadata()
+{
+    AssertLockHeld(cs_wallet); // mapKeyMetadata
+
+    for (auto& meta_pair : mapKeyMetadata) {
+        CKeyMetadata& meta = meta_pair.second;
+        if (!meta.hd_seed_id.IsNull() && meta.key_origin.IsNull() && meta.hdKeypath != "s") { // If the hdKeypath is "s", that's the seed and it doesn't have a key origin
+            CKey key;
+            GetKey(meta.hd_seed_id, key);
```

Comment:
Since this is called on load, does the user need to unlock an encrypted wallet?

Reply:
Hmm. I think it will. There is currently no way to get the master key fingerprint without generating it from the hd seed.

Reply:
I'd like to know what's going to happen here, but we should at least be checking the result here to not calculate bogus data, yes?

Reply:
I think it will write bogus data.

There are a few options to solve this. We could not upgrade the key metadata if we can't calculate the master key id. Or we could use the hd seed id instead if the correct master key id can't be calculated.

I am tending to go towards the second solution. While it technically is not the master key id, this information is really only used for/needed for descriptors and PSBT. Any wallet that already exists does not need this information. If this ends up in PSBT, that's fine since the BIP 32 derivation paths are just information for the signer. In this case, the signer is Bitcoin Core, and it doesn't use the derivation paths. We just need to put something there that won't cause problems for other signers.

Reply:
It seems there is an `IsLocked()` at the top now, so you've picked the first solution (not upgrading when locked)? If so, I prefer that.